### ISSUE TITLE
idTokenClaims undefined for Angular 17

### ISSUE THREAD

**Issue by mikaelinmeta** [AzureAD/microsoft-authentication-library-for-js#6690]
### Core Library

MSAL.js (@azure/msal-browser)

### Core Library Version

3.0.8

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

3.5.0

### Public or Confidential Client?

Public

### Description

After updating Angular from version 16 to 17, I'm encountering an issue. Although I can successfully log in and retrieve the account using `MsalService.instance.getActiveAccount()`, the `idTokenClaims` property is coming back as undefined. This is preventing access to the user's name and their claims/permissions.

Interestingly, this issue is specific to Angular 17, as it was not present in Angular 16.

### MSAL Configuration

```javascript
{
  auth: {
    clientId: 'XXXXXX',
    authority: 'XXXXXX',
    knownAuthorities: ['XXXXXX', 'XXXXXX'
    ],
    redirectUri: '/', 
  },
  cache: {
    cacheLocation: BrowserCacheLocation.LocalStorage, 
  },
}
```

### Relevant Code Snippets

```javascript
import { MsalService } from '@azure/msal-angular';

// component rendered after successful login
constructor(
    private authService: MsalService,
  ) {
    const account = this.authService.instance.getActiveAccount();
    const claims = account.idTokenClaims;
    console.log("Account: ", account, " , claims: ", claims); // claims is undefined
}
```

### Reproduction Steps

1. Update an Angular application from version 16 to 17.
2. Implement authentication using `@azure/msal-angular` and `@azure/msal-browser`.
3. Login, end the currenct session (close tab or browser), revisit the site and attempt to access `idTokenClaims` while stil logged in.

### Expected Behavior

The `idTokenClaims` property should provide access to the user's ID token claims, including their username and permissions.

### Identity Provider

Azure B2C Custom Policy

### Browsers Affected (Select all that apply)

Chrome

### Source

Internal (Microsoft)

**Comment by jingfee** [comment]
I have the same problem but I haven't upgraded to Angular 17, I'm on Angular 16. I upgraded msal-angular from 3.0.4 to 3.0.8, msal-browser from 3.1.0 to 3.5.0 and msal-common from 14.0.3 to 14.4.0 and started noticing the same issue. After downgrading all the packages to the previous versions everything works fine again.

EDIT: Did some testing on different versions and the culprit seems to be msal-browser and the issue was introduced in v3.4.0

**Comment by mikaelinmeta** [comment]
Thanks for the comment @jingfee. That's quite possible, I did bump up those versions in the same go when I upgraded to Angular 17. But I tried to downgrade to the previous versions msal-angular (3.0.4) and msal-browser (3.1.0) when I was on Angular 17 but the issue remain. Possible I did some mistake in the downgrade and the node_modules wasn't representative, I'll give it another try.

**Comment by jingfee** [comment]
> Thanks for the comment @jingfee. That's quite possible, I did bump up those versions in the same go when I upgraded to Angular 17. But I tried to downgrade to the previous versions msal-angular (3.0.4) and msal-browser (3.1.0) when I was on Angular 17 but the issue remain. Possible I did some mistake in the downgrade and the node_modules wasn't representative, I'll give it another try.

Keep in mind if you downgrade but still have the ^-prefix it won't downgrade (for example "^3.0.4" instead of "3.0.4")

**Comment by hectormmg** [comment]
@mikaelinmeta can you check if there is an ID token for that account in the cache? As of 3.4.0 idTokenClaims are only returned in the account if there is an existing ID token that matches that account in the cache.

@jingfee are you also using B2C?

**Comment by jingfee** [comment]
> @mikaelinmeta can you check if there is an ID token for that account in the cache? As of 3.4.0 idTokenClaims are only returned in the account if there is an existing ID token that matches that account in the cache.
> 
> @jingfee are you also using B2C?

Yes we are using B2C

**Comment by hectormmg** [comment]
@jingfee can you also check if the account in question has a matching ID token in the cache? Thanks!

**Comment by jingfee** [comment]
> @jingfee can you also check if the account in question has a matching ID token in the cache? Thanks!

Yes I do have a matching ID token in the cache. Double checked against msalService.instance.getAllAccounts() and there is only one account present (the one with a matching ID token in the cache) which has idTokenClaims undefined.

**Comment by hectormmg** [comment]
@jingfee as of 3.4.0 idTokenClaims are no longer cached in the account. Instead, they are added to the `AccountInfo` object returned if there is a matching ID token in the cache. This is by design.

While idTokenClaims should not always be expected in accounts returned by MSAL, in your specific scenarios (with a matching ID token) they should still be there. This might be a bug, I'll look into it.

@jingfee Can you please share MSAL trace logs of the getAllAccounts call? [Logging instructions](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/logging.md)

**Comment by hectormmg** [comment]
@jingfee @mikaelinmeta Also, does clearing the cache and then re-authenticating fix the issue?

**Comment by jingfee** [comment]
@hectormmg I'll get you a trace log tomorrow, thank you for looking into this. Re-authenticating does not fix the issue for me. Also one more thing - everything works fine on the first pageload after signing in, but when I refresh the page or try to navigate to another route the issue occurs and from what I can see by consolelogging the account the only difference is that idTokenClaims is undefined and the idToken-property is completely missing.

**Comment by mikaelinmeta** [comment]
@hectormmg I've experienced the same thing as @jingfee describes.

**Comment by jingfee** [comment]
@hectormmg 
Logfile after successfully logging in without errors
[1700034988220.log](https://github.com/AzureAD/microsoft-authentication-library-for-js/files/13361865/1700034988220.log)

Logfile after refreshing the page and getting the error
[1700035088577.log](https://github.com/AzureAD/microsoft-authentication-library-for-js/files/13361867/1700035088577.log)

**Comment by NTSAlves** [comment]
Also getting this same issue with current version of msal-react
EDIT: seems to be a custom domain issue, once reverted back to default (*.b2clogin.com) it appeared to work.
hope that helps somewhat.

**Comment by jingfee** [comment]
> Also getting this same issue with current version of msal-react EDIT: seems to be a custom domain issue, once reverted back to default (*.b2clogin.com) it appeared to work. hope that helps somewhat.

Thats not the case for me, I have the problem with the default domain.

**Comment by NikoGJ** [comment]
Also getting the same symptoms with msal-browser 3.5.0 (Using it with Vue). It did worked well with msal-browser 3.1.0

**Comment by NikoGJ** [comment]
> Also getting the same symptoms with msal-browser 3.5.0 (Using it with Vue). It did worked well with msal-browser 3.1.0

Allright, after some debugging, I found the root cause (for my case, but maybe it can help for other people who have the same issue)
It was because the environment property of my idtoken did not match any of the knownauthorities of the auth config (the config object passed when initializing the msal instance)

Details: The method CacheManager.getIdToken() wasn't able to return the IdToken from the browser cache. It was successfully able to retrieve all idtokens from the cache (which in my case, there was only one) but failed to validate it against my config. It tries to match it against the clientId, homeAccountId, environment and so on. In my case, it didn't match on the environment <-> knownauthorities because I had used the same value for 'knownauthorities' than for 'authority' and it was slighly different (I'm using ADFS)

**Comment by Devvox93** [comment]
I'm in the same situation as @jingfee. Upgraded from 
```json
"@azure/msal-angular": "^3.0.4",
"@azure/msal-browser": "^3.1.0",
```
to
```json
"@azure/msal-angular": "^3.0.8",
"@azure/msal-browser": "^3.5.0",
```

> as of 3.4.0 idTokenClaims are no longer cached in the account. Instead, they are added to the `AccountInfo` object returned if there is a matching ID token in the cache. **This is by design.**

@hectormmg  When you say that, do you mean it already worked that way and should always be used from the `AccountInfo` object? Or did it move from there in 3.4.0? That would have been a breaking change, if I correctly understand.

> This might be a bug, I'll look into it.

You also mention this is a bug, so I'm now confused if the issue will be resolved, or if we need to change our implementation.
We currently access the claim value like this:
```ts
this.claim = this.msalService.instance.getAllAccounts()[0]?.idTokenClaims["claimName"]?.toString();
```

**Comment by Core121** [comment]
We are also having this issue on my team. The idTokenClaims are valid on the initial login and once the account has to be pulled from the cache (on a refresh or similar) it is undefined.

**Comment by Core121** [comment]
> > Also getting the same symptoms with msal-browser 3.5.0 (Using it with Vue). It did worked well with msal-browser 3.1.0
> 
> Allright, after some debugging, I found the root cause (for my case, but maybe it can help for other people who have the same issue) It was because the environment property of my idtoken did not match any of the knownauthorities of the auth config (the config object passed when initializing the msal instance)
> 
> Details: The method CacheManager.getIdToken() wasn't able to return the IdToken from the browser cache. It was successfully able to retrieve all idtokens from the cache (which in my case, there was only one) but failed to validate it against my config. It tries to match it against the clientId, homeAccountId, environment and so on. In my case, it didn't match on the environment <-> knownauthorities because I had used the same value for 'knownauthorities' than for 'authority' and it was slighly different (I'm using ADFS)

I just added the environment to my list of knownAuthorities and it immediately worked. In my case I just had to lop off https:// in front where I specified the knownAuthority. Nice find!

**Comment by hectormmg** [comment]
@NikoGJ thanks for the info, as you point out the issue is MSAL not being able to match the environment because there is no authority metadata for the configured authority.

@Devvox93 what kind of authority are you using? If you're using a standard Entra ID authority then it'd be a bug, which would've been fixed in 3.6.0 which has already been released. If not, then you need to add the authority to knownAuthorities and this should be resolved like others have said.

**Comment by spaceship-zebra** [comment]
For future readers: `knownAuthorities` is supposed to be a list of **host** names. Instead of using `https://foo.example.org/bar` I've changed it to `foo.example.org` and the issue is resolved. You can also find the domain when looking in your localStorage and searching for `environment`.

Reference: https://learn.microsoft.com/en-us/dotnet/api/microsoft.authentication.webassembly.msal.msalauthenticationoptions.knownauthorities?view=aspnetcore-10.0
